vnext/Microsoft.ReactNative/Utils/BatchingEventEmitter.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "BatchingEventEmitter.h"
+#include "DynamicWriter.h"
+#include "JSValueWriter.h"
+
+namespace winrt::Microsoft::ReactNative {
+
+BatchingEventEmitter::BatchingEventEmitter(Mso::CntPtr<const Mso::React::IReactContext> &&context) noexcept
+    : m_context(std::move(context)) {
+  m_uiDispatcher = m_context->Properties().Get(ReactDispatcherHelper::UIDispatcherProperty()).as<IReactDispatcher>();
+}
+
+void BatchingEventEmitter::EmitJSEvent(int64_t tag, winrt::hstring &&eventName, JSValue &&eventObject) noexcept {
+  return EmitJSEvent(L"receiveEvent", tag, std::move(eventName), std::move(eventObject));
+}
+
+void BatchingEventEmitter::EmitJSEvent(
+    winrt::hstring &&emitterMethod,
+    int64_t tag,
+    winrt::hstring &&eventName,
+    JSValue &&eventObject) noexcept {
+  VerifyElseCrash(m_uiDispatcher.HasThreadAccess());
+
+  implementation::BatchedEvent newEvent{std::move(emitterMethod), tag, std::move(eventName), std::move(eventObject)};
+
+  size_t queueSizeAfterInsert;
+
+  {
+    std::scoped_lock lock(m_eventQueueMutex);
+
+    m_eventQueue.push_back(std::move(newEvent));
+    queueSizeAfterInsert = m_eventQueue.size();
+  }
+
+  // Once a frame is presented, send a task to JS to emit all elements in the
+  // queue. We know the task is pending on the UI thread or JS thread if we
+  // have a non-zero size queue, so we want to only trigger pumping when the
+  // first element is added to an empty queue.
+  if (queueSizeAfterInsert == 1) {
+    m_renderingRevoker = xaml::Media::CompositionTarget::Rendering(
```

Comment:
I think you have a bug here... if you have just 1 coalesced event in the queue, if you add a second coalesced event with the same key, you'll remove the 1 event in the erase-remove bit, and then you'll add a new event, re-registering for Rendering. Probably want to check if m_renderingRevoker already has a valid token in it before registering? So, like:
```
bool shouldRegister = queueWasEmpty && !m_renderingRevoker;
```
(I think revokers have an `operator bool` for this?)

Reply:
Yes, it looks like there is a bug here. We cannot rely on the revoker though, just because it is possible we have an open batch with unregistered revoker. Seems like we can check the size before coalescing though.